### PR TITLE
feat(time-zone-picker): allow custom defaults

### PR DESCRIPTION
**Pull request by gavin-everett-genesys**
Open to opinions on this change https://inindca.atlassian.net/browse/COMUI-2381

It was hard to come up with a solution that seemed good. My suggested approach works well but I am unsure. The `work-space-default` and `local-default` properties seem to be use case specific so I did not want to alter them. The only logical solution I could think of was to add a `custom-default` property which the end user would set along with a label.

I know this component is still in beta so maybe we could alter it a bit more.

Once we agree on an approach this feeback will be sent to the UX team to be reflected in the designs.

Any feedback is welcome :)

### CODE REVIEW

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/example.html
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/example.html:11]: Do we have any validation on these strings to check if they are really timezones?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:27]: Could you add some annotation to indicate the difference between workspace default and custom default?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:123]: Remove

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:36]: What's the difference between the `customDefaultLabel` prop and `customDefault`?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:173]: i wonder if we can have the word `zone` in the `workspaceDefault` and `localDefault` props to signify that they are related to zones?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:36]: The main difference between the two is that for example if you were to choose `localDefault` or `workspaceDefault` only one property is needed as the the string is hardcoded to either `Local Default` or `Workspace Default` and you just pass your time zone to the property whereas with the  two custom properties we would need 2 because I do not know what label the consumer will provide so we would need two properties one for the label and one for the timezone which they want to use. It would be nice to use property somehow but I dont think this is possible.

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:173]: I think this would be a breaking change if I were to change the property names.

**Comment by gavin-everett-genesys** [comment]
Since this PR has gone stale I want to provide an update on the changes made.

So we currently have `workSpaceDefault` and `localDefault` which are use-case specific. For example if you use the `workSpaceDefault` property you will specify a timezone and this will appear under the defaults section in the dropdown with the text hardcoded to `( Workspace Default ) `.

My change alters the API a small bit but does not cause a breaking change. I have introduced two new properties `customDefault` and `customDefaultLabel`. The `customDefaultLabel` property will allow the consumer to provide a label for example " (BU Timezone) " as outlined in the ticket. And then `customDefault` would be to specify a timezone.

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:23]: How come this has changed to mutable? I don't see any other change to the value here?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:36]: Gotcha, thanks!

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:172]: Ah nevermind

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:155]: what does this line do? `].filter(zone => !!zone);`? For my own understanding. Are you trying to filter out empty values from workspaceDefault, localDefault, and customDefault?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:179]: The zone list is user defined, right? Wondering if the user would ever have an extra space or even a typo in the props: defaultZone and workspaceDefault and then the exact match check here would fail? Maybe that's not our problem and would be on the user?

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:155]: Ye so basically if a user has provided a timezone for those properties it will filter those out out say for example you have `gux-timezone-picker-beta` setup like,

```
  <gux-time-zone-picker-beta
    value="Etc/GMT+1"
    workspace-default="Etc/GMT"
    local-default="America/Detroit"
  > 
```

The funtion would then filter out `Etc/GMT` and `America/Detroit` and store those in an array which will be used to display under the default zones list.

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:179]: Thats a good question I am not sure as of yet as there was no validation incorporated for the initial properties which were `workspace-default` and `local-default`. I will wait till we are happy with this api change and review this.

**Comment by 321gillian** [comment]
Tests failing, might just need to update snapshot

File: packages/genesys-spark-components/src/components/beta/gux-date-time/tests/__snapshots__/gux-date-time.spec.ts.snap
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-date-time/tests/__snapshots__/gux-date-time.spec.ts.snap:334]: I think the failure is due to this not sure why it was picked up but will investigate it.

**Review by 321gillian — APPROVED** [review]

**Comment by jason-evans-genesys** [comment]
@gavin-everett-genesys I noticed there's no keyboard navigation way to get to the default zone area in the time-zone-picker dropdown. Is this expected?

Screen Shot 2024-06-17 at 10 21 11 AM

**Review by katie-bobbe-genesys — CHANGES_REQUESTED** [review]
As Jason mentioned, the default zones do not seem to be keyboard accessible. We should resolve this issue before merging

**Comment by gavin-everett-genesys** [comment]
> @gavin-everett-genesys I noticed there's no keyboard navigation way to get to the default zone area in the time-zone-picker dropdown. Is this expected?
> 
> Screen Shot 2024-06-17 at 10 21 11 AM

I think this exists in the live version aswell. I will work on a fix in this PR 👍

**Comment by gavin-everett-genesys** [comment]
> As Jason mentioned, the default zones do not seem to be keyboard accessible. We should resolve this issue before merging

Updated to use `gux-option-group-beta` to display `gux-option` elements.

File: packages/genesys-spark-components/src/components/stable/gux-icon/readme.md
Comment (jordanstith15) [inline packages/genesys-spark-components/src/components/stable/gux-icon/readme.md:13]: With this many icons, maybe we should discuss whether the automatic script for the icon name attribute is worth it. Feels like it just makes the readme less readable

**Review by jordanstith15 — APPROVED** [review]

File: packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx
Comment (jason-evans-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-time-zone-picker/gux-time-zone-picker.tsx:205]: For some reason having a type of JSX.Element or null feels off to me here. I feel like it should be only one type. Is there a way to render an empty JSX.Element maybe instead of null?

**Review by katie-bobbe-genesys — COMMENTED** [review]
This looks good to me. Just make sure to resolve conflicts in the icons readme before merging.

**Review by katie-bobbe-genesys — APPROVED** [review]

**Pull request merged**
